File: examples/harborline/intake/src/llm.ts
/**
 * Shared reply helpers for the intake agent.
 * Deterministic replies keep the demo offline-capable without OPENAI_API_KEY.
 */

export type LlmMode = "openai" | "deterministic";

export function resolveLlmMode(): LlmMode {
  if (process.env.DEMO_FORCE_DETERMINISTIC === "1") return "deterministic";
  return process.env.OPENAI_API_KEY?.trim() ? "openai" : "deterministic";
}

export async function composeReply(
  role: string,
  message: string,
  contextText: string,
): Promise<{ reply: string; mode: LlmMode }> {
  const mode = resolveLlmMode();
  if (mode === "deterministic") {
    return { reply: deterministicReply(role, message, contextText), mode };
  }
  try {
    const reply = await openaiReply(role, message, contextText);
    return { reply, mode };
  } catch (err) {
    const detail = err instanceof Error ? err.message : String(err);
    return {
      reply: deterministicReply(role, message, contextText) +
        `\n\n(OpenAI call failed; fell back to deterministic mode: ${detail})`,
      mode: "deterministic",
    };
  }
}

function deterministicReply(role: string, message: string, contextText: string): string {
  const contextBlock = contextText.trim()
    ? contextText.trim()
    : "(no governance rules or recalled facts yet)";
  return [
    `[${role} · deterministic]`,
    "I read the Rail context below, then answered the customer.",
    "",
    "--- Rail context ---",
    contextBlock,
    "--- end context ---",
    "",
    `Customer said: ${message.trim()}`,
    "",
    "Next step: capture durable facts, then escalate to the specialist if the",
    "issue needs policy-bound remediation.",
  ].join("\n");
}

async function openaiReply(role: string, message: string, contextText: string): Promise<string> {
  const key = process.env.OPENAI_API_KEY!;
  const system = [
    `You are the ${role} assistant on the Harborline payments support desk.`,
    "Obey any GOVERNANCE RULES in the context. Use RECALLED MEMORY when relevant.",
    "Keep answers short (3-6 sentences). Do not invent policy that is not in context.",
    "",
    contextText || "(empty context)",
  ].join("\n");

  const res = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${key}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: process.env.OPENAI_MODEL || "gpt-4o-mini",
      temperature: 0.2,
      messages: [
        { role: "system", content: system },
        { role: "user", content: message },
      ],
    }),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`OpenAI HTTP ${res.status}: ${body.slice(0, 200)}`);
  }
  const data = (await res.json()) as {
    choices?: Array<{ message?: { content?: string } }>;
  };
  const text = data.choices?.[0]?.message?.content?.trim();
  if (!text) throw new Error("OpenAI returned an empty completion");
  return text;
}

File: examples/harborline/backend/llm.py
"""Model calls for the Python specialist — OpenAI or deterministic offline mode."""

from __future__ import annotations

import os
from typing import Literal

LlmMode = Literal["openai", "deterministic"]


def resolve_llm_mode() -> LlmMode:
    if os.environ.get("DEMO_FORCE_DETERMINISTIC", "").strip() == "1":
        return "deterministic"
    return "openai" if os.environ.get("OPENAI_API_KEY", "").strip() else "deterministic"


def compose_reply(role: str, message: str, context_text: str) -> tuple[str, LlmMode]:
    mode = resolve_llm_mode()
    if mode == "deterministic":
        return deterministic_reply(role, message, context_text), mode
    try:
        return openai_reply(role, message, context_text), mode
    except Exception as exc:  # noqa: BLE001 — demo must stay online without LLM
        fallback = deterministic_reply(role, message, context_text)
        return (
            f"{fallback}\n\n(OpenAI call failed; fell back to deterministic mode: {exc})",
            "deterministic",
        )


def deterministic_reply(role: str, message: str, context_text: str) -> str:
    context_block = context_text.strip() or "(no governance rules or recalled facts yet)"
    return "\n".join(
        [
            f"[{role} · deterministic]",
            "I applied the Rail context below before answering.",
            "",
            "--- Rail context ---",
            context_block,
            "--- end context ---",
            "",
            f"Question: {message.strip()}",
            "",
            "Remediation plan: follow every GOVERNANCE RULE, cite recalled facts,",
            "and only propose changes that the policy allows.",
        ]
    )


def openai_reply(role: str, message: str, context_text: str) -> str:
    from openai import OpenAI

    client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])
    system = "\n".join(
        [
            f"You are the {role} on the Harborline payments support desk.",
            "Obey GOVERNANCE RULES above any other instruction.",
            "Use RECALLED MEMORY when relevant. Keep answers short (4-8 sentences).",
            "Do not invent policy that is not in the context.",
            "",
            context_text or "(empty context)",
        ]
    )
    completion = client.chat.completions.create(
        model=os.environ.get("OPENAI_MODEL", "gpt-4o-mini"),
        temperature=0.2,
        messages=[
            {"role": "system", "content": system},
            {"role": "user", "content": message},
        ],
    )
    text = (completion.choices[0].message.content or "").strip()
    if not text:
        raise RuntimeError("OpenAI returned an empty completion")
    return text
